core/chains/evm/gas/models.go

```diff
@@ -261,32 +264,53 @@ func (e *evmFeeEstimator) L1Oracle() rollups.L1Oracle {
 	return e.EvmEstimator.L1Oracle()
 }
 
-func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, opts ...feetypes.Opt) (fee EvmFee, chainSpecificFeeLimit uint64, err error) {
+func (e *evmFeeEstimator) GetFee(ctx context.Context, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, toAddress *common.Address, opts ...feetypes.Opt) (fee EvmFee, chainSpecificFeeLimit uint64, err error) {
+	// Create call msg for gas limit estimation, if needed
+	callMsg := ethereum.CallMsg{
+		To:   toAddress,
+		Data: calldata,
+	}
+
 	// get dynamic fee
 	if e.EIP1559Enabled {
 		var dynamicFee DynamicFee
 		dynamicFee, err = e.EvmEstimator.GetDynamicFee(ctx, maxFeePrice)
 		if err != nil {
 			return
 		}
-		chainSpecificFeeLimit, err = commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
 		fee.DynamicFeeCap = dynamicFee.FeeCap
 		fee.DynamicTipCap = dynamicFee.TipCap
-		return
+		chainSpecificFeeLimit = feeLimit
+		// Set call msg fields with dynamic fee fields for gas limit estimation, if needed
+		callMsg.GasFeeCap = fee.DynamicFeeCap.ToInt()
+		callMsg.GasTipCap = fee.DynamicTipCap.ToInt()
+	} else {
+		// get legacy fee
+		fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
+		if err != nil {
+			return
+		}
+		// Set call msg fields with legacy fee field for gas limit estimation, if needed
+		callMsg.GasPrice = fee.Legacy.ToInt()
 	}
 
-	// get legacy fee
-	fee.Legacy, chainSpecificFeeLimit, err = e.EvmEstimator.GetLegacyGas(ctx, calldata, feeLimit, maxFeePrice, opts...)
-	if err != nil {
-		return
+	if e.geCfg.EstimateGasLimit() {
```

Comment:
For readability, could you create a new function, called getEstimatedGasUsed()?
And the ethereum.CallMsg, .geCfg.EstimateGasLimit() checks, client calls, please move all to that function.
This makes it very clear to reader what part is for the GetFee(), vs for getEstimatedGasUsed().